File: src/services/storageService.ts
/**
 * Storage Service - Unified abstraction for S3 and Local filesystem storage
 */
import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import config from '../config';
import logger from '../utils/logger';
import { s3Service } from './s3Service';

const mkdir = promisify(fs.mkdir);
const writeFile = promisify(fs.writeFile);
const readFile = promisify(fs.readFile);
const unlink = promisify(fs.unlink);
const stat = promisify(fs.stat);

// =============================================================================
// Storage Provider Interface
// =============================================================================

export interface IStorageProvider {
  uploadFile(key: string, body: Buffer | string, contentType: string): Promise<{ key: string; url: string }>;
  uploadPDF(workspaceId: string, fileName: string, buffer: Buffer, userEmail?: string): Promise<{ key: string; url: string }>;
  uploadText(workspaceId: string, title: string, content: string, userEmail?: string): Promise<{ key: string; url: string }>;
  getSignedReadUrl(key: string, expiresIn?: number): Promise<string>;
  deleteFile(key: string): Promise<void>;
  getFileContent(key: string): Promise<string>;
  getFileBuffer(key: string): Promise<Buffer>;
  fileExists(key: string): Promise<boolean>;
}

// =============================================================================
// Local Storage Provider
// =============================================================================

class LocalStorageProvider implements IStorageProvider {
  private basePath: string;

  constructor(basePath?: string) {
    this.basePath = path.resolve(basePath || config.localStoragePath);
    this.ensureBaseDir();
    logger.info(`[LocalStorage] Initialized with base path: ${this.basePath}`);
  }

  private ensureBaseDir(): void {
    if (!fs.existsSync(this.basePath)) {
      fs.mkdirSync(this.basePath, { recursive: true });
    }
  }

  private getFullPath(key: string): string {
    // Normalize the key to prevent path traversal
    const safeKey = key.replace(/\.\./g, '').replace(/^\/+/, '');
    return path.join(this.basePath, safeKey);
  }

  private async ensureDir(filePath: string): Promise<void> {
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) {
      await mkdir(dir, { recursive: true });
    }
  }

  async uploadFile(
    key: string,
    body: Buffer | string,
    contentType: string
  ): Promise<{ key: string; url: string; contentType: string }> {
    const filePath = this.getFullPath(key);
    await this.ensureDir(filePath);

    if (typeof body === 'string') {
      await writeFile(filePath, body, 'utf-8');
    } else {
      await writeFile(filePath, body);
    }

    const publicUrl = await this.getSignedReadUrl(key);
    logger.info(`[LocalStorage] Saved file: ${filePath}`);
    
    return {
      key,
      url: publicUrl,
      contentType,
    };
  }

  async uploadPDF(
    workspaceId: string,
    fileName: string,
    buffer: Buffer,
    userEmail?: string
  ): Promise<{ key: string; url: string }> {
    const key = `${userEmail || 'anonymous'}/workspaces/${workspaceId}/pdfs/${Date.now()}-${fileName}`;
    const result = await this.uploadFile(key, buffer, 'application/pdf');
    return { key: result.key, url: result.url };
  }

  async uploadText(
    workspaceId: string,
    title: string,
    content: string,
    userEmail?: string
  ): Promise<{ key: string; url: string }> {
    const sanitizedTitle = title.replace(/[^a-zA-Z0-9]/g, '_');
    const key = `${userEmail || 'anonymous'}/workspaces/${workspaceId}/texts/${Date.now()}-${sanitizedTitle}.txt`;
    const result = await this.uploadFile(key, content, 'text/plain');
    return { key: result.key, url: result.url };
  }

  async getSignedReadUrl(key: string, _expiresIn = 3600): Promise<string> {
    // Return HTTP URL pointing to our static file server
    // Use the backend URL (default to localhost:4000 for development)
    const baseUrl = process.env.BACKEND_URL || `http://localhost:${config.port}`;
    return `${baseUrl}/storage/${key}`;
  }

  async deleteFile(key: string): Promise<void> {
    const filePath = this.getFullPath(key);
    if (fs.existsSync(filePath)) {
      await unlink(filePath);
      logger.info(`[LocalStorage] Deleted file: ${filePath}`);
    }
  }

  async getFileContent(key: string): Promise<string> {
    const filePath = this.getFullPath(key);
    return await readFile(filePath, 'utf-8');
  }

  async getFileBuffer(key: string): Promise<Buffer> {
    const filePath = this.getFullPath(key);
    return await readFile(filePath);
  }

  async fileExists(key: string): Promise<boolean> {
    const filePath = this.getFullPath(key);
    try {
      await stat(filePath);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Get the absolute local path for direct file access
   */
  getLocalPath(key: string): string {
    return this.getFullPath(key);
  }
}

// =============================================================================
// S3 Storage Provider (Wrapper around existing s3Service)
// =============================================================================

class S3StorageProvider implements IStorageProvider {
  constructor() {
    logger.info(`[S3Storage] Initialized with bucket: ${config.aws.s3Bucket}`);
  }

  async uploadFile(
    key: string,
    body: Buffer | string,
    contentType: string
  ): Promise<{ key: string; url: string }> {
    return s3Service.uploadFile(key, body, contentType);
  }

  async uploadPDF(
    workspaceId: string,
    fileName: string,
    buffer: Buffer,
    userEmail?: string
  ): Promise<{ key: string; url: string }> {
    return s3Service.uploadPDF(workspaceId, fileName, buffer, userEmail);
  }

  async uploadText(
    workspaceId: string,
    title: string,
    content: string,
    userEmail?: string
  ): Promise<{ key: string; url: string }> {
    return s3Service.uploadText(workspaceId, title, content, userEmail);
  }

  async getSignedReadUrl(key: string, expiresIn = 3600): Promise<string> {
    return s3Service.getSignedReadUrl(key, expiresIn);
  }

  async deleteFile(key: string): Promise<void> {
    return s3Service.deleteFile(key);
  }

  async getFileContent(key: string): Promise<string> {
    return s3Service.getFileContent(key);
  }

  async getFileBuffer(key: string): Promise<Buffer> {
    const content = await s3Service.getFileContent(key);
    return Buffer.from(content);
  }

  async fileExists(_key: string): Promise<boolean> {
    // S3 doesn't have a simple exists check, would need HeadObject
    // For now, attempt getFileContent and catch errors
    try {
      await this.getFileContent(_key);
      return true;
    } catch {
      return false;
    }
  }
}

// =============================================================================
// Factory Function
// =============================================================================

let _storageProvider: IStorageProvider | null = null;

/**
 * Get the configured storage provider (S3 or Local)
 * 
 * @example
 * ```typescript
 * const storage = getStorageProvider();
 * await storage.uploadPDF(workspaceId, 'doc.pdf', buffer, userEmail);
 * ```
 */
export function getStorageProvider(): IStorageProvider {
  if (_storageProvider === null) {
    const provider = config.storageProvider.toLowerCase();

    if (provider === 's3') {
      _storageProvider = new S3StorageProvider();
    } else if (provider === 'local') {
      _storageProvider = new LocalStorageProvider();
    } else {
      throw new Error(`Unknown storage provider: ${provider}. Available: s3, local`);
    }

    logger.info(`[Storage] Using ${provider.toUpperCase()} storage provider`);
  }

  return _storageProvider;
}

/**
 * Convenience alias for getStorageProvider
 */
export const storageService = {
  get provider(): IStorageProvider {
    return getStorageProvider();
  },

  // Proxy methods for convenience
  async uploadFile(key: string, body: Buffer | string, contentType: string) {
    return this.provider.uploadFile(key, body, contentType);
  },

  async uploadPDF(workspaceId: string, fileName: string, buffer: Buffer, userEmail?: string) {
    return this.provider.uploadPDF(workspaceId, fileName, buffer, userEmail);
  },

  async uploadText(workspaceId: string, title: string, content: string, userEmail?: string) {
    return this.provider.uploadText(workspaceId, title, content, userEmail);
  },

  async getSignedReadUrl(key: string, expiresIn = 3600) {
    return this.provider.getSignedReadUrl(key, expiresIn);
  },

  async deleteFile(key: string) {
    return this.provider.deleteFile(key);
  },

  async getFileContent(key: string) {
    return this.provider.getFileContent(key);
  },

  async getFileBuffer(key: string) {
    return this.provider.getFileBuffer(key);
  },

  async fileExists(key: string) {
    return this.provider.fileExists(key);
  },
};

export default storageService;

File: ai/app/services/storage_service.py
# Storage Service - Abstraction for S3 and Local filesystem storage
import os
import shutil
from abc import ABC, abstractmethod
from typing import Optional
from pathlib import Path
from app.config import settings


class BaseStorageProvider(ABC):
    """Abstract base class for storage providers."""
    
    @abstractmethod
    async def upload_file(self, key: str, content: bytes, content_type: str = "application/octet-stream") -> str:
        """Upload a file and return the URL/path."""
        pass
    
    @abstractmethod
    async def get_file_content(self, key: str) -> bytes:
        """Get file content by key."""
        pass
    
    @abstractmethod
    async def delete_file(self, key: str) -> bool:
        """Delete a file by key."""
        pass
    
    @abstractmethod
    async def get_file_url(self, key: str, expires_in: int = 3600) -> str:
        """Get a URL to access the file (signed URL for S3, local path for local)."""
        pass
    
    @abstractmethod
    def file_exists(self, key: str) -> bool:
        """Check if a file exists."""
        pass


class LocalStorageProvider(BaseStorageProvider):
    """Local filesystem storage provider."""
    
    def __init__(self, base_path: str = None):
        self.base_path = Path(base_path or settings.LOCAL_STORAGE_PATH)
        # Ensure the base directory exists
        self.base_path.mkdir(parents=True, exist_ok=True)
        print(f"[LocalStorage] Initialized with base path: {self.base_path.absolute()}")
    
    def _get_full_path(self, key: str) -> Path:
        """Get the full filesystem path for a key."""
        # Normalize the key to prevent path traversal
        safe_key = key.lstrip("/").replace("..", "")
        return self.base_path / safe_key
    
    async def upload_file(self, key: str, content: bytes, content_type: str = "application/octet-stream") -> str:
        """Upload a file to local storage."""
        file_path = self._get_full_path(key)
        
        # Create parent directories if they don't exist
        file_path.parent.mkdir(parents=True, exist_ok=True)
        
        # Write the file
        if isinstance(content, str):
            file_path.write_text(content, encoding='utf-8')
        else:
            file_path.write_bytes(content)
        
        print(f"[LocalStorage] Saved file: {file_path}")
        return str(file_path.absolute())
    
    async def get_file_content(self, key: str) -> bytes:
        """Get file content from local storage."""
        file_path = self._get_full_path(key)
        
        if not file_path.exists():
            raise FileNotFoundError(f"File not found: {key}")
        
        return file_path.read_bytes()
    
    async def delete_file(self, key: str) -> bool:
        """Delete a file from local storage."""
        file_path = self._get_full_path(key)
        
        if file_path.exists():
            file_path.unlink()
            print(f"[LocalStorage] Deleted file: {file_path}")
            return True
        return False
    
    async def get_file_url(self, key: str, expires_in: int = 3600) -> str:
        """Get the local file path as URL."""
        file_path = self._get_full_path(key)
        # Return as file:// URL for local access
        return f"file://{file_path.absolute()}"
    
    def file_exists(self, key: str) -> bool:
        """Check if a file exists."""
        return self._get_full_path(key).exists()
    
    def get_local_path(self, key: str) -> str:
        """Get the absolute local path for direct file access."""
        return str(self._get_full_path(key).absolute())


class S3StorageProvider(BaseStorageProvider):
    """AWS S3 storage provider."""
    
    def __init__(self):
        import boto3
        
        self.bucket = settings.AWS_S3_BUCKET
        self.region = settings.AWS_REGION
        
        # Initialize S3 client
        client_kwargs = {'region_name': self.region}
        
        # Add credentials if provided
        if settings.AWS_ACCESS_KEY_ID and settings.AWS_SECRET_ACCESS_KEY:
            client_kwargs['aws_access_key_id'] = settings.AWS_ACCESS_KEY_ID
            client_kwargs['aws_secret_access_key'] = settings.AWS_SECRET_ACCESS_KEY
        
        self.s3 = boto3.client('s3', **client_kwargs)
        print(f"[S3Storage] Initialized with bucket: {self.bucket}")
    
    async def upload_file(self, key: str, content: bytes, content_type: str = "application/octet-stream") -> str:
        """Upload a file to S3."""
        self.s3.put_object(
            Bucket=self.bucket,
            Key=key,
            Body=content,
            ContentType=content_type
        )
        
        url = f"https://{self.bucket}.s3.{self.region}.amazonaws.com/{key}"
        print(f"[S3Storage] Uploaded file: {key}")
        return url
    
    async def get_file_content(self, key: str) -> bytes:
        """Get file content from S3."""
        response = self.s3.get_object(Bucket=self.bucket, Key=key)
        return response['Body'].read()
    
    async def delete_file(self, key: str) -> bool:
        """Delete a file from S3."""
        try:
            self.s3.delete_object(Bucket=self.bucket, Key=key)
            print(f"[S3Storage] Deleted file: {key}")
            return True
        except Exception as e:
            print(f"[S3Storage] Error deleting file {key}: {e}")
            return False
    
    async def get_file_url(self, key: str, expires_in: int = 3600) -> str:
        """Get a pre-signed URL for S3 object."""
        url = self.s3.generate_presigned_url(
            'get_object',
            Params={'Bucket': self.bucket, 'Key': key},
            ExpiresIn=expires_in
        )
        return url
    
    def file_exists(self, key: str) -> bool:
        """Check if a file exists in S3."""
        try:
            self.s3.head_object(Bucket=self.bucket, Key=key)
            return True
        except:
            return False


# =============================================================================
# Factory Function
# =============================================================================

_storage_provider: Optional[BaseStorageProvider] = None


def get_storage_provider() -> BaseStorageProvider:
    """
    Factory function to get the configured storage provider.
    
    Returns:
        Configured storage provider instance (S3 or Local)
    
    Example:
        >>> storage = get_storage_provider()
        >>> await storage.upload_file("docs/file.pdf", content, "application/pdf")
    """
    global _storage_provider
    
    if _storage_provider is None:
        provider_type = settings.STORAGE_PROVIDER.lower()
        
        if provider_type == "s3":
            _storage_provider = S3StorageProvider()
        elif provider_type == "local":
            _storage_provider = LocalStorageProvider()
        else:
            raise ValueError(
                f"Unknown storage provider: {provider_type}. "
                f"Available: s3, local"
            )
        
        print(f"[Storage] Using {provider_type.upper()} storage provider")
    
    return _storage_provider


# Default storage provider instance (lazy loaded)
def get_storage() -> BaseStorageProvider:
    """Convenience alias for get_storage_provider()."""
    return get_storage_provider()
